Moq.AutoMock/ResolutionContext.cs

```diff
@@ -0,0 +1,40 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Moq.AutoMock
+{
+    /// <summary>
+    /// Handles state while creating an object graph.
+    /// </summary>
+    public class ResolutionContext
+    {
+        /// <summary>
+        /// Creates an instance with binding flags set according to `enablePrivate`.
+        /// </summary>
+        /// <param name="enablePrivate"></param>
+        public ResolutionContext(bool enablePrivate)
+        {
+            BindingFlags = GetBindingFlags(enablePrivate);
+            VisitedTypes = new HashSet<Type>();
+        }
+
+        /// <summary>
+        /// Flags passed to Mock constructor.
+        /// </summary>
+        public BindingFlags BindingFlags { get; }
+        
+        /// <summary>
+        /// Used internally to track which types have been created inside a call graph,
+        /// to detect cycles in the object graph.
+        /// </summary>
+        public HashSet<Type> VisitedTypes { get; }
```

Comment:
Since object graphs are not likely to be very large. I would suggest making this a list so that the exception message can properly display the hierarchy with the recursive dependency.